meerk40t/gui/scene/scene.py

```diff
@@ -141,6 +144,20 @@ def signal(self, *args, **kwargs):
         """
         self.scene._signal_widget(self.scene.widget_root, *args, **kwargs)
 
+    def on_key_down(self, evt):
+        keycode = evt.GetKeyCode()
+        if keycode == wx.WXK_SHIFT:
+            self.scene.isShiftPressed = True
+        if keycode == wx.WXK_CONTROL:
+            self.scene.isCtrlPressed = True
+
+    def on_key_up(self, evt):
+        keycode = evt.GetKeyCode()
+        if keycode == wx.WXK_SHIFT:
+            self.scene.isShiftPressed = False
+        if keycode == wx.WXK_CONTROL:
+            self.scene.isCtrlPressed = False
```

Comment:
My gut reaction is that this is **NOT** the way to go. IMO the raw event that wx executes should be queried to find out what modifiers are pressed **at the time of the leftup event** (and every other event) and these modifiers should be passed as parameters to the various classes that processes the event. There is also a routine in wxutil which you should use to check for modifiers because there are some variances between o/s and this routine handles these.

Reply:
Also, I think that Ctrl-drag and Shift-drag and Ctrl-Shift-drag are all valid actions, and we should decide what each of these does. I have a feeling that Shift is usually additive, Control is usually toggle.

Reply:
I get your reaction to the treatment of Ctrl & Shift - as stated that has been a quick and dirty attempt to see whether this is beneficial or not (and after having used now for some hours I personally like it a lot) and needs to be refined. It's definitely not up to MK code standards yet.